lib/Mediator.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Amp;
+
+/**
+ * A Promise/Coroutine-aware implementation of the Mediator pattern
+ * 
+ * @link https://en.wikipedia.org/wiki/Mediator_pattern
+ */
+class Mediator
+{
+    private $eventSubscriberMap = [];
+
+    /**
+     * Invoke the callback when the specified event is published to the Mediator
+     * 
+     * If a subscriber callback returns FALSE (===) or a generator\promise that
+     * resolves to FALSE that callback will be unsubscribed from the event.
+     * 
+     * @param string $eventName The name of the event being subscribed to
+     * @param callable $callback The callback to invoke when the event is published
+     * @return void
+     */
+    public function subscribe(string $eventName, callable $callback)
+    {
+        $this->eventSubscriberMap[$eventName][] = $callback;
+    }
+
+    /**
+     * Publish an event with associated data payload for consumption by subscribers
+     * 
+     * If a subscriber callback returns FALSE (===) or a generator\promise that
+     * resolves to FALSE that callback will be unsubscribed from the event.
+     * 
+     * @param string $eventName The name of the event to publish
+     * @param mixed $data Data associated with the event
+     * @return \Amp\Promise<int> Resolves to the number of subscribers invoked
+     */
+    public function publish(string $eventName, ...$data): Promise
+    {
+        if (empty($this->eventSubscriberMap[$eventName])) {
+            return new Success(0);
+        }
+        return new Coroutine((function() use ($eventName, $data) {
+            $errors = [];
+            $executionCount = 0;
+            $callbacks = $this->eventSubscriberMap[$eventName];
+            foreach ($callbacks as $id => $callback) {
+                try {
+                    // detach callback if it returns false (===)
+                    if (false === yield call($callback, ...$data)) {
```

Comment:
Any reason why these are executed sequentially?

Reply:
General programming rust on my part -- didn't even think about doing it concurrently. Will update.